File: src/main/java/edu/metrostate/ics372/ganby/vehicle/Vehicle.java
package edu.metrostate.ics372.ganby.vehicle;

import edu.metrostate.ics372.ganby.dealer.Dealer;
import javafx.beans.property.*;
import java.time.LocalDateTime;
import java.util.Objects;

/**
 * Abstract base class representing a vehicle.
 * Provides shared attributes and behaviors for all vehicle types such as ID, model, price, acquisition date, and rental status.
 * Supports JavaFX property bindings for UI integration.
 */
public abstract class Vehicle {

    // Properties for JavaFX bindings
    protected final StringProperty id;
    protected final StringProperty model;
    protected final StringProperty manufacturer;
    protected final DoubleProperty price;
    protected final StringProperty dealerId;
    protected final ObjectProperty<LocalDateTime> acquisitionDate;
    protected final BooleanProperty isRentedOut;
    private final BooleanProperty selected = new SimpleBooleanProperty(false);
    // --- JavaFX Bindable Properties ---

    public StringProperty vehicleIdProperty() {
        return id;
    }

    public StringProperty modelProperty() {
        return model;
    }

    public StringProperty manufacturerProperty() {
        return manufacturer;
    }

    public DoubleProperty priceProperty() {
        return price;
    }

    public StringProperty dealerIdProperty() {
        return dealerId;
    }

    public ObjectProperty<LocalDateTime> acquisitionDateProperty() {
        return acquisitionDate;
    }

    public BooleanProperty isRentedOutProperty() {
        return isRentedOut;
    }

    /**
     * Constructs a vehicle with rental status.
     *
     * @param id              the vehicle ID
     * @param model           the vehicle model
     * @param manufacturer    the manufacturer of the vehicle
     * @param price           the price of the vehicle
     * @param dealerId        the ID of the associated dealer
     * @param acquisitionDate the date the vehicle was acquired
     * @param isRentedOut     rental status of the vehicle
     */
    public Vehicle(String id, String model, String manufacturer, double price, String dealerId,
                   LocalDateTime acquisitionDate, Boolean isRentedOut) {
        this.id = new SimpleStringProperty(id);
        this.model = new SimpleStringProperty(model);
        this.manufacturer = new SimpleStringProperty(manufacturer);
        this.price = new SimpleDoubleProperty(price);
        this.dealerId = new SimpleStringProperty(dealerId);
        this.acquisitionDate = new SimpleObjectProperty<>(acquisitionDate);
        this.isRentedOut = new SimpleBooleanProperty(isRentedOut);
    }

    /**
     * Constructs a vehicle with rental status defaulted to false.
     *
     * @param id              the vehicle ID
     * @param model           the vehicle model
     * @param manufacturer    the manufacturer of the vehicle
     * @param price           the price of the vehicle
     * @param dealerId        the ID of the associated dealer
     * @param acquisitionDate the date the vehicle was acquired
     */
    public Vehicle(String id, String model, String manufacturer, double price, String dealerId,
                   LocalDateTime acquisitionDate) {
        this(id, model, manufacturer, price, dealerId, acquisitionDate, false);
    }

    /**
     * Returns the vehicle type (e.g., SUV, Sedan).
     *
     * @return the vehicle type
     */
    public abstract String getType();

    // --- Getters ---

    /**
     * @return the vehicle ID
     */
    public String getVehicleId() {
        return id.get();
    }

    /**
     * @return the vehicle model
     */
    public String getModel() {
        return model.get();
    }

    /**
     * @return the vehicle manufacturer
     */
    public String getManufacturer() {
        return manufacturer.get();
    }

    /**
     * @return the vehicle price
     */
    public double getPrice() {
        return price.get();
    }

    /**
     * @return the dealer ID this vehicle belongs to
     */
    public String getDealerId() {
        return dealerId.get();
    }

    /**
     * @return the acquisition date of the vehicle
     */
    public LocalDateTime getAcquisitionDate() {
        return acquisitionDate.get();
    }

    /**
     * @return true if rented out, false otherwise
     */
    public Boolean getIsRentedOut() {
        return isRentedOut.get();
    }

    // --- Setters ---

    /**
     * Sets the price of the vehicle.
     *
     * @param price new price (must be non-negative)
     */
    public void setPrice(double price) {
        if (price < 0) throw new IllegalArgumentException("Price cannot be negative");
        this.price.set(price);
    }

    /**
     * Sets the dealer this vehicle belongs to.
     *
     * @param dealer the dealer instance
     */
    public void setDealer(Dealer dealer) {
        if (dealer == null) throw new IllegalArgumentException("Dealer cannot be null");
        this.dealerId.set(dealer.getId());
    }

    /**
     * Sets the dealer ID directly.
     *
     * @param dealerId dealer's ID
     */
    public void setDealerId(String dealerId) {
        this.dealerId.set(dealerId);
    }

    /**
     * Sets the acquisition date of the vehicle.
     *
     * @param newAcquisitionDate the new acquisition date
     */
    public void setAcquisitionDate(LocalDateTime newAcquisitionDate) {
        if (newAcquisitionDate == null)
            throw new IllegalArgumentException("New acquisition date cannot be null");
        this.acquisitionDate.set(newAcquisitionDate);
    }

    /**
     * Sets the rented-out status of the vehicle.
     * Override this method in subclasses if rental is not allowed.
     *
     * @param rentedOut the new rent status
     * @return the vehicle type (can be used to restrict rental for certain types)
     */
    public void setRentedOut(Boolean rentedOut) {
        this.isRentedOut.set(rentedOut);
    }

    // --- Selection Support ---

    /**
     * Gets the selected property for checkbox bindings.
     *
     * @return the selected property
     */
    public BooleanProperty selectedProperty() {
        return selected;
    }

    /**
     * @return true if the vehicle is selected
     */
    public boolean isSelected() {
        return selected.get();
    }

    /**
     * Sets the selected state of the vehicle (used in UI).
     *
     * @param selected true if selected, false otherwise
     */
    public void setSelected(boolean selected) {
        this.selected.set(selected);
    }


    public void VehicleBuilderFromType(String type){

    }

    // --- Object Overrides ---

    @Override
    public boolean equals(Object object) {
        if (object == this) return true;
        if (object == null || !(object instanceof Vehicle vehicle)) return false;
        return id.get().equals(vehicle.getVehicleId());
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id.get());
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() +
                " [id:" + getVehicleId()
                + " model:" + getModel()
                + " manufacturer:" + getManufacturer()
                + " price:" + getPrice()
                + " dealer id:" + getDealerId()
                + " acquisitionDate:" + getAcquisitionDate() + "]";
    }
}

File: src/main/java/edu/metrostate/ics372/ganby/vehicle/Vehicle.kt
package edu.metrostate.ics372.ganby.vehicle

import edu.metrostate.ics372.ganby.dealer.Dealer
import javafx.beans.property.*
import java.time.LocalDateTime
import java.util.*

/**
 * Abstract base class representing a vehicle.
 *
 * This class provides shared attributes and behaviors for all vehicle types, such as:
 * - Vehicle ID
 * - Model
 * - Manufacturer
 * - Price
 * - Dealer association
 * - Acquisition date
 * - Rental status
 * - UI selection support
 *
 * It is designed for use with JavaFX applications by exposing JavaFX properties for all fields.
 *
 * @constructor Creates a new vehicle instance with required fields and optional rental flag.
 * @param id Unique vehicle identifier.
 * @param model Model name of the vehicle.
 * @param manufacturer Manufacturer name.
 * @param price Price of the vehicle (non-negative).
 * @param dealerId ID of the associated dealer.
 * @param acquisitionDate Date the vehicle was acquired.
 * @param isRentedOut Boolean flag indicating if the vehicle is currently rented out.
 */
abstract class Vehicle @JvmOverloads constructor(
    id: String,
    model: String,
    manufacturer: String,
    price: Double,
    dealerId: String,
    acquisitionDate: LocalDateTime,
    isRentedOut: Boolean = false
) {
    // JavaFX property bindings for all fields
    private val _id = SimpleStringProperty(id)
    private val _model = SimpleStringProperty(model)
    private val _manufacturer = SimpleStringProperty(manufacturer)
    private val _price = SimpleDoubleProperty(price)
    private val _dealerId = SimpleStringProperty(dealerId)
    private val _acquisitionDate = SimpleObjectProperty(acquisitionDate)
    private val _isRentedOut = SimpleBooleanProperty(isRentedOut)
    private val _selected = SimpleBooleanProperty(false)

    /**
     * Abstract vehicle type to be implemented by subclasses (e.g., "SUV", "Sedan").
     */
    abstract val type: String

    // ----- JavaFX Property Getters (for GUI bindings) -----

    fun idProperty(): StringProperty = _id
    fun modelProperty(): StringProperty = _model
    fun manufacturerProperty(): StringProperty = _manufacturer
    fun priceProperty(): DoubleProperty = _price
    fun dealerIdProperty(): StringProperty = _dealerId
    fun acquisitionDateProperty(): ObjectProperty<LocalDateTime> = _acquisitionDate
    fun isRentedOutProperty(): BooleanProperty = _isRentedOut
    fun selectedProperty(): BooleanProperty = _selected

    // ----- Standard Getters -----

    /** @return the vehicle's unique ID */
    fun getVehicleId(): String = _id.get()

    /** @return the vehicle's model name */
    fun getModel(): String = _model.get()

    /** @return the vehicle's manufacturer */
    fun getManufacturer(): String = _manufacturer.get()

    /** @return the vehicle's current price */
    fun getPrice(): Double = _price.get()

    /** @return the ID of the dealer who owns this vehicle */
    fun getDealerId(): String = _dealerId.get()

    /** @return the acquisition date */
    fun getAcquisitionDate(): LocalDateTime = _acquisitionDate.get()

    /** @return whether the vehicle is currently rented out */
    fun getIsRentedOut(): Boolean = _isRentedOut.get()

    /** @return whether the vehicle is selected in the UI */
    fun isSelected(): Boolean = _selected.get()

    // ----- Setters -----

    /**
     * Updates the vehicle's price.
     * @param newPrice the new price (must be non-negative)
     * @throws IllegalArgumentException if newPrice < 0
     */
    fun setPrice(newPrice: Double) {
        require(newPrice >= 0) { "Price cannot be negative" }
        _price.set(newPrice)
    }

    /**
     * Sets the dealer for this vehicle and updates the dealer ID.
     * @param dealer the dealer object (must not be null)
     * @throws IllegalArgumentException if dealer is null
     */
    fun setDealer(dealer: Dealer) {
        dealer?.let {
            _dealerId.set(it.getId())
        } ?: throw IllegalArgumentException("Dealer cannot be null")
    }

    /**
     * Updates the dealer ID for this vehicle.
     * @param newDealerId new dealer ID
     */
    fun setDealerId(newDealerId: String) {
        _dealerId.set(newDealerId)
    }

    /**
     * Updates the acquisition date of the vehicle.
     * @param newDate the new acquisition date (non-null)
     * @throws IllegalArgumentException if newDate is null
     */
    fun setAcquisitionDate(newDate: LocalDateTime) {
        requireNotNull(newDate) { "New acquisition date cannot be null" }
        _acquisitionDate.set(newDate)
    }

    /**
     * Updates rental status.
     * @param rentedOut true if the vehicle is currently rented out
     */
    fun setRentedOut(rentedOut: Boolean) {
        _isRentedOut.set(rentedOut)
    }

    /**
     * Updates UI selection status for table checkboxes.
     * @param isSelected true to mark as selected
     */
    fun setSelected(isSelected: Boolean) {
        _selected.set(isSelected)
    }

    // ----- Object Overrides -----

    /** Two vehicles are equal if they have the same ID. */
    override fun equals(other: Any?): Boolean {
        return other is Vehicle && _id.get() == other._id.get()
    }

    override fun hashCode(): Int = Objects.hash(_id.get())

    override fun toString(): String {
        return "${javaClass.simpleName} [id: ${getVehicleId()}, model: ${getModel()}, manufacturer: ${getManufacturer()}, price: ${getPrice()}, dealerId: ${getDealerId()}, acquisitionDate: ${getAcquisitionDate()}]"
    }
}
